File: libs/graphicsgeo/src/main/kotlin/mono/graphics.geo/Rect.kt
package mono.graphics.geo

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

@Serializable(with = Rect.RectSerializer::class)
data class Rect(
    val position: Point,
    val size: Size
) {
    val width: Int = size.width
    val height: Int = size.height

    val left: Int = position.left
    val right: Int = position.left + width - 1
    val top: Int = position.top
    val bottom: Int = position.top + height - 1

    private val validHorizontalRange = left..right
    private val validVerticalRange = top..bottom

    operator fun contains(point: Point): Boolean =
        point.left in validHorizontalRange && point.top in validVerticalRange

    fun getOverlappedRect(rect: Rect): Rect? {
        if (!isOverlapped(rect)) {
            return null
        }
        val offset = rect.position - position
        val top = max(offset.top, 0)
        val bottom = min(offset.top + rect.height, height) - 1
        val left = max(offset.left, 0)
        val right = min(offset.left + rect.width, width) - 1
        return byLTRB(
            left + position.left,
            top + position.top,
            right + position.left,
            bottom + position.top
        )
    }

    fun isOverlapped(rect: Rect): Boolean {
        val isHorizontalOverlap = left in rect.left..rect.right || rect.left in left..right
        val isVerticalOverlap = top in rect.top..rect.bottom || rect.top in top..bottom
        return isHorizontalOverlap && isVerticalOverlap
    }

    override fun toString(): String = "[$left, $top] - [$width x $height]"

    internal object RectSerializer : KSerializer<Rect> {
        override val descriptor: SerialDescriptor =
            PrimitiveSerialDescriptor("Rect", PrimitiveKind.STRING)

        override fun serialize(encoder: Encoder, value: Rect) {
            encoder.encodeString("${value.left}|${value.top}|${value.width}|${value.height}")
        }

        override fun deserialize(decoder: Decoder): Rect {
            val marshaledValue = decoder.decodeString()
            val (left, top, width, height) = marshaledValue.split("|")
            return byLTWH(left.toInt(), top.toInt(), width.toInt(), height.toInt())
        }
    }

    companion object {
        val ZERO = byLTWH(0, 0, 0, 0)

        fun byLTRB(left: Int, top: Int, right: Int, bottom: Int): Rect = Rect(
            Point(min(left, right), min(top, bottom)),
            Size(abs(right - left) + 1, abs(bottom - top) + 1)
        )

        fun byLTWH(left: Int, top: Int, width: Int, height: Int): Rect = Rect(
            Point(left, top),
            Size(width, height)
        )

        fun boundOf(rectangles: Sequence<Rect>): Rect {
            val firstRect = rectangles.first()
            var left = firstRect.left
            var right = firstRect.right
            var top = firstRect.top
            var bottom = firstRect.bottom

            for (rect in rectangles) {
                left = min(rect.left, left)
                right = max(rect.right, right)
                top = min(rect.top, top)
                bottom = max(rect.bottom, bottom)
            }
            return byLTRB(left, top, right, bottom)
        }
    }
}

File: monosketch-svelte/src/lib/libs/graphics-geo/rect.ts
import { type IPoint, Point } from './point';
import { Size } from './size';
import { IntRange } from '$libs/sequence';
import type { Comparable } from '$libs/comparable';

export class Rect implements Comparable {
    static readonly ZERO = Rect.byLTWH(0, 0, 0, 0);

    static byLTRB(left: number, top: number, right: number, bottom: number): Rect {
        [left, right] = [Math.min(left, right), Math.max(left, right)];
        [top, bottom] = [Math.min(top, bottom), Math.max(top, bottom)];
        return new Rect(Point.of(left, top), Size.of(right - left + 1, bottom - top + 1));
    }

    static byLTRBf(left: number, top: number, right: number, bottom: number): Rect {
        [left, right] = [Math.min(left, right), Math.max(left, right)];
        [top, bottom] = [Math.min(top, bottom), Math.max(top, bottom)];
        return new Rect(Point.ofF(left, top), Size.ofF(right - left + 1, bottom - top + 1));
    }

    static byLTWH(left: number, top: number, width: number, height: number): Rect {
        return new Rect(Point.of(left, top), Size.of(width, height));
    }

    static byLTWHf(left: number, top: number, width: number, height: number): Rect {
        return new Rect(Point.ofF(left, top), Size.ofF(width, height));
    }

    private readonly validHorizontalRange: IntRange;
    private readonly validVerticalRange: IntRange;

    constructor(
        public readonly position: Point,
        public readonly size: Size,
    ) {
        this.validHorizontalRange = new IntRange(this.left, this.right + 1);
        this.validVerticalRange = new IntRange(this.top, this.bottom + 1);
    }

    get width(): number {
        return this.size.width;
    }

    get height(): number {
        return this.size.height;
    }

    get left(): number {
        return this.position.left;
    }

    get top(): number {
        return this.position.top;
    }

    get right(): number {
        return this.left + this.width - 1;
    }

    get bottom(): number {
        return this.top + this.height - 1;
    }

    equals(other: unknown): boolean {
        if (!(other instanceof Rect)) {
            return false;
        }
        return this.position.equals(other.position) && this.size.equals(other.size);
    }

    contains(point: IPoint): boolean {
        return (
            this.validHorizontalRange.contains(point.left) &&
            this.validVerticalRange.contains(point.top)
        );
    }

    isOverlapped(other: Rect): boolean {
        const isHorizontalOverlapped =
            this.validHorizontalRange.contains(other.left) ||
            other.validHorizontalRange.contains(this.left);
        const isVerticalOverlapped =
            this.validVerticalRange.contains(other.top) ||
            other.validVerticalRange.contains(this.top);

        return isHorizontalOverlapped && isVerticalOverlapped;
    }

    getOverlappedRect(rect: Rect): Rect | null {
        if (!this.isOverlapped(rect)) {
            return null;
        }
        const offset = rect.position.minus(this.position);
        const top = Math.max(offset.top, 0);
        const bottom = Math.min(offset.top + rect.height, this.height) - 1;
        const left = Math.max(offset.left, 0);
        const right = Math.min(offset.left + rect.width, this.width) - 1;
        return Rect.byLTRB(
            left + this.position.left,
            top + this.position.top,
            right + this.position.left,
            bottom + this.position.top,
        );
    }

    /**
     * Returns true if the point is one of its vertices.
     */
    isVertex(point: Point): boolean {
        const isHorizontalVertex = point.left === this.left || point.left === this.right;
        const isVerticalVertex = point.top === this.top || point.top === this.bottom;
        return isHorizontalVertex && isVerticalVertex;
    }

    toString(): string {
        return `[${this.left}, ${this.top}] - [${this.width} x ${this.height}]`;
    }

    // TODO: implement serialize and deserialize
}
